File: backend/app/strategies/formula_evaluator.py
"""安全公式表达式求值（禁止 eval）。"""

from __future__ import annotations

import re
from dataclasses import dataclass
from decimal import Decimal, InvalidOperation
from enum import Enum
from typing import Callable

REF_PATTERN = re.compile(
    r"^@(?P<ind>[a-zA-Z_][a-zA-Z0-9_]*)(?:\.(?P<out>[a-zA-Z_][a-zA-Z0-9_]*))?$"
    r"|^\$(?P<field>open|high|low|close|volume)$"
    r"|^#(?P<param>[a-zA-Z_][a-zA-Z0-9_]*)$"
    r"|^&(?P<formula>[a-zA-Z_][a-zA-Z0-9_]*)$"
)


class FormulaError(Exception):
    pass


def _safe_decimal(value) -> Decimal | None:
    if value is None:
        return None
    try:
        d = Decimal(str(value)) if not isinstance(value, Decimal) else value
    except (InvalidOperation, ValueError, TypeError):
        return None
    if not d.is_finite():
        return None
    return d


class TokKind(Enum):
    NUMBER = "number"
    REF = "ref"
    PLUS = "+"
    MINUS = "-"
    STAR = "*"
    SLASH = "/"
    LPAREN = "("
    RPAREN = ")"
    EOF = "eof"


@dataclass
class Token:
    kind: TokKind
    value: str = ""


def tokenize(expr: str) -> list[Token]:
    expr = expr.strip()
    if not expr:
        raise FormulaError("公式不能为空")
    tokens: list[Token] = []
    i = 0
    while i < len(expr):
        ch = expr[i]
        if ch.isspace():
            i += 1
            continue
        if ch in "+*/()":
            tokens.append(Token(TokKind(ch), ch))
            i += 1
            continue
        if ch == "-":
            tokens.append(Token(TokKind.MINUS, "-"))
            i += 1
            continue
        if ch in "@$#&":
            j = i
            while j < len(expr) and (expr[j].isalnum() or expr[j] in "._$#@&"):
                j += 1
            tokens.append(Token(TokKind.REF, expr[i:j]))
            i = j
            continue
        if ch.isdigit() or ch == ".":
            j = i
            while j < len(expr) and (expr[j].isdigit() or expr[j] == "."):
                j += 1
            tokens.append(Token(TokKind.NUMBER, expr[i:j]))
            i = j
            continue
        raise FormulaError(f"非法字符: {ch}")
    tokens.append(Token(TokKind.EOF))
    return tokens


class _Evaluator:
    def __init__(self, tokens: list[Token], resolver: Callable[[str], Decimal | None]):
        self._tokens = tokens
        self._pos = 0
        self._resolve = resolver

    def _cur(self) -> Token:
        return self._tokens[self._pos]

    def _eat(self, kind: TokKind | None = None) -> Token:
        tok = self._cur()
        if kind and tok.kind != kind:
            raise FormulaError(f"语法错误，期望 {kind.value}，实际 {tok.kind.value}")
        self._pos += 1
        return tok

    def evaluate(self) -> Decimal | None:
        val = self._expr()
        if self._cur().kind != TokKind.EOF:
            raise FormulaError("表达式未预期结束")
        return val

    def _expr(self) -> Decimal | None:
        val = self._term()
        while self._cur().kind in (TokKind.PLUS, TokKind.MINUS):
            op = self._eat().kind
            rhs = self._term()
            if val is None or rhs is None:
                return None
            val = val + rhs if op == TokKind.PLUS else val - rhs
        return val

    def _term(self) -> Decimal | None:
        val = self._factor()
        while self._cur().kind in (TokKind.STAR, TokKind.SLASH):
            op = self._eat().kind
            rhs = self._factor()
            if val is None or rhs is None:
                return None
            if op == TokKind.STAR:
                val = val * rhs
            else:
                if rhs == 0:
                    return None
                val = val / rhs
        return val

    def _factor(self) -> Decimal | None:
        if self._cur().kind == TokKind.MINUS:
            self._eat(TokKind.MINUS)
            val = self._factor()
            return -val if val is not None else None
        if self._cur().kind == TokKind.NUMBER:
            return _safe_decimal(self._eat(TokKind.NUMBER).value)
        if self._cur().kind == TokKind.REF:
            return self._resolve(self._eat(TokKind.REF).value)
        if self._cur().kind == TokKind.LPAREN:
            self._eat(TokKind.LPAREN)
            val = self._expr()
            self._eat(TokKind.RPAREN)
            return val
        raise FormulaError(f"语法错误: {self._cur().kind.value}")


def evaluate_expression(expression: str, resolver: Callable[[str], Decimal | None]) -> Decimal | None:
    tokens = tokenize(expression)
    return _Evaluator(tokens, resolver).evaluate()


class FormulaEngine:
    """按拓扑顺序求值公式列表。"""

    def __init__(
        self,
        *,
        formulas: list[dict],
        indicators: dict,
        parameters: dict,
        bar_fields: dict[str, Decimal],
        use_prev: bool = False,
    ):
        self._formula_defs = [
            f for f in formulas if isinstance(f, dict) and f.get("id") and f.get("expression")
        ]
        self._indicators = indicators
        self._parameters = parameters
        self._bar_fields = bar_fields
        self._use_prev = use_prev
        self._values: dict[str, Decimal | None] = {}
        self._prev_values: dict[str, Decimal | None] = {}

    def _base_resolver(self, ref: str, values: dict[str, Decimal | None]) -> Decimal | None:
        m = REF_PATTERN.match(ref)
        if not m:
            raise FormulaError(f"非法引用: {ref}")
        if m.group("formula"):
            return values.get(m.group("formula"))
        if m.group("ind"):
            ind = self._indicators.get(m.group("ind"))
            if ind is None:
                return None
            output = m.group("out") or "value"
            if self._use_prev:
                return _safe_decimal(ind.get_prev_output(output))
            return _safe_decimal(ind.get_output(output))
        if m.group("field"):
            key = f"_prev_{m.group('field')}" if self._use_prev else m.group("field")
            return _safe_decimal(self._bar_fields.get(key))
        if m.group("param"):
            return _safe_decimal(self._parameters.get(m.group("param")))
        return None

    def evaluate_all(self) -> dict[str, Decimal | None]:
        ordered = self._topo_sort()
        current: dict[str, Decimal | None] = {}
        for fid in ordered:
            expr = next(f["expression"] for f in self._formula_defs if f["id"] == fid)
            current[fid] = evaluate_expression(
                expr, lambda ref, vals=current: self._base_resolver(ref, vals)
            )
        self._prev_values = dict(self._values)
        self._values = current
        return current

    def get(self, formula_id: str) -> Decimal | None:
        return self._values.get(formula_id)

    def get_prev(self, formula_id: str) -> Decimal | None:
        return self._prev_values.get(formula_id)

    def _topo_sort(self) -> list[str]:
        ids = [f["id"] for f in self._formula_defs]
        deps: dict[str, set[str]] = {}
        for f in self._formula_defs:
            fid = f["id"]
            refs: set[str] = set()
            for tok in tokenize(f["expression"]):
                if tok.kind == TokKind.REF and tok.value.startswith("&"):
                    refs.add(tok.value[1:])
            deps[fid] = refs
        ordered: list[str] = []
        temp: set[str] = set()

        def visit(n: str) -> None:
            if n in temp:
                raise FormulaError(f"公式循环引用: {n}")
            if n in ordered:
                return
            temp.add(n)
            for d in deps.get(n, set()):
                if d in ids:
                    visit(d)
            temp.remove(n)
            ordered.append(n)

        for fid in ids:
            visit(fid)
        return ordered

File: frontend/src/utils/formulaEvaluator.ts
/** 安全公式求值（与后端 formula_evaluator.py 对齐，禁止 eval）。 */

export type FormulaPreviewContext = {
  indicators: Record<string, number | Record<string, number>>;
  parameters: Record<string, number>;
  barFields: Record<string, number>;
  formulas: Record<string, number | null>;
};

export class FormulaError extends Error {
  constructor(message: string) {
    super(message);
    this.name = "FormulaError";
  }
}

const TokKind = {
  NUMBER: "number",
  REF: "ref",
  PLUS: "+",
  MINUS: "-",
  STAR: "*",
  SLASH: "/",
  LPAREN: "(",
  RPAREN: ")",
  EOF: "eof",
} as const;

type TokKind = (typeof TokKind)[keyof typeof TokKind];

type Token = { kind: TokKind; value: string };

function tokenize(expr: string): Token[] {
  const trimmed = expr.trim();
  if (!trimmed) throw new FormulaError("公式不能为空");
  const tokens: Token[] = [];
  let i = 0;
  while (i < trimmed.length) {
    const ch = trimmed[i];
    if (/\s/.test(ch)) {
      i += 1;
      continue;
    }
    if ("+*/()".includes(ch)) {
      tokens.push({ kind: ch as TokKind, value: ch });
      i += 1;
      continue;
    }
    if (ch === "-") {
      tokens.push({ kind: TokKind.MINUS, value: "-" });
      i += 1;
      continue;
    }
    if ("@$#&".includes(ch)) {
      let j = i;
      while (j < trimmed.length && /[\w.$#@&]/.test(trimmed[j])) j += 1;
      tokens.push({ kind: TokKind.REF, value: trimmed.slice(i, j) });
      i = j;
      continue;
    }
    if (/[\d.]/.test(ch)) {
      let j = i;
      while (j < trimmed.length && /[\d.]/.test(trimmed[j])) j += 1;
      tokens.push({ kind: TokKind.NUMBER, value: trimmed.slice(i, j) });
      i = j;
      continue;
    }
    throw new FormulaError(`非法字符: ${ch}`);
  }
  tokens.push({ kind: TokKind.EOF, value: "" });
  return tokens;
}

function resolveRef(ref: string, ctx: FormulaPreviewContext): number | null {
  if (ref.startsWith("&")) {
    const id = ref.slice(1);
    const val = ctx.formulas[id];
    return val === undefined || val === null ? null : val;
  }
  if (ref.startsWith("@")) {
    const body = ref.slice(1);
    const dot = body.indexOf(".");
    const id = dot >= 0 ? body.slice(0, dot) : body;
    const output = dot >= 0 ? body.slice(dot + 1) : "value";
    const ind = ctx.indicators[id];
    if (ind === undefined) return null;
    if (typeof ind === "number") return output === "value" ? ind : null;
    const val = ind[output];
    return val === undefined ? null : val;
  }
  if (ref.startsWith("$")) {
    const field = ref.slice(1);
    const val = ctx.barFields[field];
    return val === undefined ? null : val;
  }
  if (ref.startsWith("#")) {
    const param = ref.slice(1);
    const val = ctx.parameters[param];
    return val === undefined ? null : val;
  }
  return null;
}

class Evaluator {
  private pos = 0;
  private tokens: Token[];
  private resolver: (ref: string) => number | null;

  constructor(tokens: Token[], resolver: (ref: string) => number | null) {
    this.tokens = tokens;
    this.resolver = resolver;
  }

  evaluate(): number | null {
    const val = this.expr();
    if (this.cur().kind !== TokKind.EOF) throw new FormulaError("表达式未预期结束");
    return val;
  }

  private cur(): Token {
    return this.tokens[this.pos];
  }

  private eat(kind?: TokKind): Token {
    const tok = this.cur();
    if (kind && tok.kind !== kind) {
      throw new FormulaError(`语法错误，期望 ${kind}，实际 ${tok.kind}`);
    }
    this.pos += 1;
    return tok;
  }

  private expr(): number | null {
    let val = this.term();
    while (this.cur().kind === TokKind.PLUS || this.cur().kind === TokKind.MINUS) {
      const op = this.eat().kind;
      const rhs = this.term();
      if (val === null || rhs === null) return null;
      val = op === TokKind.PLUS ? val + rhs : val - rhs;
    }
    return val;
  }

  private term(): number | null {
    let val = this.factor();
    while (this.cur().kind === TokKind.STAR || this.cur().kind === TokKind.SLASH) {
      const op = this.eat().kind;
      const rhs = this.factor();
      if (val === null || rhs === null) return null;
      if (op === TokKind.STAR) val = val * rhs;
      else {
        if (rhs === 0) return null;
        val = val / rhs;
      }
    }
    return val;
  }

  private factor(): number | null {
    if (this.cur().kind === TokKind.MINUS) {
      this.eat(TokKind.MINUS);
      const val = this.factor();
      return val === null ? null : -val;
    }
    if (this.cur().kind === TokKind.NUMBER) {
      const n = Number(this.eat(TokKind.NUMBER).value);
      return Number.isFinite(n) ? n : null;
    }
    if (this.cur().kind === TokKind.REF) {
      return this.resolver(this.eat(TokKind.REF).value);
    }
    if (this.cur().kind === TokKind.LPAREN) {
      this.eat(TokKind.LPAREN);
      const val = this.expr();
      this.eat(TokKind.RPAREN);
      return val;
    }
    throw new FormulaError(`语法错误: ${this.cur().kind}`);
  }
}

export function evaluateExpression(
  expression: string,
  resolver: (ref: string) => number | null,
): number | null {
  const tokens = tokenize(expression);
  return new Evaluator(tokens, resolver).evaluate();
}

function topoSort(formulas: Array<{ id: string; expression: string }>): string[] {
  const ids = formulas.map((f) => f.id);
  const deps: Record<string, Set<string>> = {};
  for (const f of formulas) {
    const refs = new Set<string>();
    for (const tok of tokenize(f.expression)) {
      if (tok.kind === TokKind.REF && tok.value.startsWith("&")) {
        refs.add(tok.value.slice(1));
      }
    }
    deps[f.id] = refs;
  }
  const ordered: string[] = [];
  const temp = new Set<string>();

  const visit = (n: string) => {
    if (temp.has(n)) throw new FormulaError(`公式循环引用: ${n}`);
    if (ordered.includes(n)) return;
    temp.add(n);
    for (const d of deps[n] ?? []) {
      if (ids.includes(d)) visit(d);
    }
    temp.delete(n);
    ordered.push(n);
  };

  for (const id of ids) visit(id);
  return ordered;
}

export function evaluateFormulas(
  formulas: Array<{ id: string; expression: string }>,
  ctx: Omit<FormulaPreviewContext, "formulas">,
): { values: Record<string, number | null>; error?: string } {
  const values: Record<string, number | null> = {};
  try {
    const ordered = topoSort(formulas);
    for (const id of ordered) {
      const expr = formulas.find((f) => f.id === id)!.expression;
      values[id] = evaluateExpression(expr, (ref) =>
        resolveRef(ref, { ...ctx, formulas: values }),
      );
    }
    return { values };
  } catch (err) {
    return {
      values,
      error: err instanceof Error ? err.message : String(err),
    };
  }
}

/** 预览用示例上下文（基于当前指标/参数生成占位值）。 */
export function buildSamplePreviewContext(
  indicatorIds: string[],
  indicatorOutputs: Record<string, string[]>,
  parameters: Record<string, unknown>,
): Omit<FormulaPreviewContext, "formulas"> {
  const indicators: Record<string, number | Record<string, number>> = {};
  indicatorIds.forEach((id, idx) => {
    const outputs = indicatorOutputs[id] ?? ["value"];
    if (outputs.length === 1) {
      indicators[id] = 10 + idx * 2;
    } else {
      const obj: Record<string, number> = {};
      outputs.forEach((out, oi) => {
        obj[out] = 10 + idx * 2 + oi;
      });
      indicators[id] = obj;
    }
  });
  const params: Record<string, number> = {};
  for (const [k, v] of Object.entries(parameters)) {
    if (typeof v === "object" && v !== null && "default" in (v as object)) {
      params[k] = Number((v as { default: unknown }).default);
    } else if (typeof v === "number") {
      params[k] = v;
    }
  }
  return {
    indicators,
    parameters: params,
    barFields: { close: 12, open: 11, high: 13, low: 10, volume: 10000 },
  };
}
